app/jobs/order_cycle_opened_job.rb

```diff
@@ -5,6 +5,39 @@ class OrderCycleOpenedJob < ApplicationJob
   def perform
     ActiveRecord::Base.transaction do
       recently_opened_order_cycles.find_each do |order_cycle|
+        # Process the order cycle. this might take a while, so this should be shifted to a separate job to allow concurrent processing.        
+
+        order_cycle.suppliers.each do |supplier|
+          # Find authorised user to access remote products
+          dfc_user = supplier.owner # we assume the owner's account is linked with dfc. maybe instead we could search for supplier.users.joins(:oidc_account).first
+
+          variants = order_cycle.exchanges.incoming.from_enterprise(supplier).joins(:exchange_variants).pluck('exchange_variants.variant_id') #todo: actually we probably only need to look at outgoing products. no need to sync products if they're not being sold.
+          links = SemanticLink.where(subject: variants) #todo: we should be able to join with above query. in fact why not includes(:subject)
```

Comment:
Have a look at the BackorderUpdater:

```rb
  def distributed_linked_variants(order_cycle, distributor)
    order_cycle.variants_distributed_by(distributor)
      .where.associated(:semantic_links)
  end
```

So you can probably simplify this to something like this:

```rb
links = SemanticLink.where(subject: order_cycle.variants_distributed_by(distributor))
```

Reply:
@mkllnk in this context, I am looping through suppliers, and not considering distributors.

There's (currently) no `OrderCycle#variants_supplied_by` method. I could probably make one, but I think this case is a bit more specific anyway.

Reply:
Hmm the good thing about that method is that it produces a Relation, rather than an array of IDs.
I had assumed that the Relation would have been resolved and queried, then fed in to the `subject:` parameter. But  I just did a  little reading and it looks like ActiveRecord can actually  handle this and actually combines it as a subquery. Clever!

Reply:
> I am looping through suppliers, and not considering distributors.

Of course! Sorry for that distraction. And it's good that you don't even go into the variants table. You could filter by outgoing exchanges, too, but I'm not sure it's worth it.